### ISSUE TITLE
MD051 creates false positives when a heading contains html comments

### ISSUE THREAD

**Issue by ITWARRIORS-Kilian** [DavidAnson/markdownlint#2206]
---

**Title:** `MD051` incorrectly includes HTML comment content in heading fragment generation

---

**What did you do?**

Used an HTML comment inside a Markdown heading:

```markdown
## Hello <!-- world -->

[link](#hello)
```

**What did you expect to happen?**

The fragment `#hello` should be considered valid, matching GitHub's anchor generation which strips HTML comments before building the slug.

**What actually happened?**

MD051 reports a violation because the generated fragment is `#hello--world--` instead of `#hello`. The comment text leaks into the slug.

**How can the issue be reproduced?**

```markdown
## Hello <!-- world -->

[link](#hello)
```

Run markdownlint against this file. MD051 fires on the `[link](#hello)` reference.

**Assumption / How this may be fixed**

In `md051.mjs`, `convertHeadingToHTMLFragment` uses `filterByPredicate` with a `childrenExclude` set. `htmlText` tokens (which cover HTML comments) are not excluded, so their `data` children are traversed and included in the slug. Adding `"htmlText"` to `childrenExclude` aligns the behavior with GitHub's anchor generation.

Expected fragment: `#hello`  
Actual fragment: `#hello--world--`

**Version**

VS code extension version 0.62.1

**Operating system**

VSCode on Windows

**Comment by ITWARRIORS-Kilian** [comment]
I can test the fix tomorrow.

**Comment by DavidAnson** [comment]
I'll fix this later today, thanks.

FYI, GitHub produces "#hello-" for your example, not "#hello".

**Comment by DavidAnson** [comment]
The current behavior is already correct. GitHub produces `#hello-` for the example above and that is accepted by MD051/link-fragments. You can see that here: <https://dlaa.me/markdownlint/#%25m%23%23%20Hello%20%3C!--%20world%20--%3E%0A%0A%5Blink%5D(%23hello-)%0A>.

**Comment by ITWARRIORS-Kilian** [comment]
Hi  @DavidAnson ,

Thank you very much for your effort!  

I tested it further and can confirm that `#hello-` indeed works on GitHub. In fact, GitHub appears to support both variations: `#hello-` and `#hello`.  

Here’s a simple test case to demonstrate this behavior (I hope this renders correctly here):  

```md
# Hallo again <!--Comment-->

[Link `#Hallo-again`](#Hallo-again)  
[Link `#Hallo-again-`](#Hallo-again-)  
```

# Hallo again 

[Link `#Hallo-again`](#Hallo-again)  
[Link `#Hallo-again-`](#Hallo-again-)  

However, in VSCode, Markdownlint only accepts the version without the trailing `-` (`#Hallo-again`). Using the trailing `-` (`#Hallo-again-`) generates a warning. For better compatibility, I propose supporting both versions.  

Additionally, I encountered another (admittedly rare) edge case:  

```md
## Hallo [Link ](www.github.com)

Note the whitespace behind the link!

[Link hallo Link no trailing](#hallo-link)  
[Link hallo Link trailing](#hallo-link-)  
```

In this example, both versions of the link (`#hallo-link` and `#hallo-link-`) work correctly on GitHub. However, in VSCode, only the version without the trailing `-` is accepted. The working version however triggers MD051.  

Thank you for considering these observations!  

Kind regards,  
Kilian

**Comment by DavidAnson** [comment]
GitHub is somewhat flexible about what they support, but there is a single format they generate and that's what this rule enforces.

When you refer to VS Code behavior, please be specific. The `markdownlint` extension (published by me) uses exactly this library and therefore has the same behavior.

Other extensions and VS Code itself may behave differently, but I cannot change them.

**Comment by ITWARRIORS-Kilian** [comment]
Hi @DavidAnson,

First, I apologize for the imprecise wording in my previous post — I incorrectly stated that markdownlint was rejecting the trailing `-` variant, when in fact the issue is simply that the link does not navigate correctly in VS Code's Markdown preview. That was misleading.  (Asking AI to improve readability doesn't always improve readability :grimacing:)

After investigating further, here is what I verified:

For the heading:
```markdown
## Hallo again <!--Comment-->
```

Using the **VS Code Webview Developer Tools** and **browser developer tools on GitHub.com and GitLab.com**, I inspected the actual rendered `id` attributes:

| Platform | Generated `id` |
|---|---|
| VS Code preview (`markdown-it` + `slugify.ts`) | `hallo-again` |
| GitHub.com | `hallo-again-` or `hallo-again` |
| GitLab.com | `hallo-again-` |

The root cause appears to be a difference between VS Code's slug algorithm (`slugify.ts` in `markdown-language-features`) and GitHub's pipeline:
- VS Code **strips** HTML comments entirely before slugifying → no trailing `-`
- GitHub **replaces** the whitespace first before stripping HTML comments → trailing `-` remains

Therefore, MD051 correctly aligns with GitHub's behavior, accepting `#hallo-again-` and flagging `#hallo-again`. I agree with you now that MD051 should stay as it is.

Since anchor generation is deliberately not part of the de facto Markdown specification (CommonMark), which all three platforms aim to align to, I sadly don't see this ambiguity in anchor generation changing in the near future. This leaves VS Code users with the dilemma that their markdown links either conform and correctly navigate in VS Code, or on GitHub/GitLab, but not both.

I recognize this is likely a rare edge case that most users won't encounter, but I wanted to raise whether it might be worth a note in MD051's documentation, or if it might be worth a new rule enforcing users to simply remove whitespace before trailing HTML tags that are stripped during anchor generation.

The proposed rule should flag cases like:

```markdown
## Hallo again <!--Comment-->
```

And enforce changing them to:

```markdown
## Hallo again<!--Comment-->
```

I'll defer to your judgment on whether it's worth addressing. If you think this could be worth the effort, I'm also willing to help implement it.

Kind regards,
Kilian

**Comment by DavidAnson** [comment]
Thank you for the extra information! I'll probably add a note about this to the documentation in a few days.
